lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+  outs() << "Struct definition: " << *StructureType << "\n";
+  for (unsigned i = 0; i < SubRecords.size(); i++) {
+    outs() << "Group #" << i + 1 << ":{ ";
+    for (auto &F : *SubRecords[i]) {
+      outs() << "F" << F + 1 << " ";
+    }
+    outs() << "}\n";
+  }
+}
+
+// Functions of OldFieldReorderAnalyzer. Here "old" just means the algorithm
+// used to reorder fields is derived from prior work in GCC and not necessarily
+// mean it's deprecated or worse compared to the other FieldReorderAnalyzer
+OldFieldReorderAnalyzer::OldFieldReorderAnalyzer(
+    const Module &CM, const StructType *ST, const CloseProximityBuilder *CPB,
+    const DICompositeType *DI)
+    : FieldReorderAnalyzer(CM, ST, CPB, DI, true),
+      DistanceThreshold(StructSplitDistanceThreshold) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Initializing OldFieldReorderAnalyzer of size: "
+                         << NumElements << "\n");
+  // Calculate every CP relations and save it to an array to avoid repeated
+  // calculation
+  CloseProximityRelations.resize(NumElements);
+  double MaxWCP = 0;
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+      if (CloseProximityRelations[i][j] > MaxWCP)
+        MaxWCP = CloseProximityRelations[i][j];
+    }
+  }
+  if (MaxWCP < 1e-3) {
+    Eligibility = false;
+    return;
+  }
+  DEBUG(dbgs() << "CPG for field reordering recommendation\n");
+  for (unsigned i = 0; i < NumElements; i++) {
+    DEBUG(dbgs() << "F" << i + 1 << "\t");
+    for (unsigned j = 0; j < NumElements; j++)
+      DEBUG(dbgs() << DEBUG_PRINT_COUNT(CloseProximityRelations[i][j]) << "\t");
+    DEBUG(dbgs() << "\n");
+  }
+}
+
+// Function to calculate CP values. In this algorithm, it interprets (count,
+// distance) as using count as it's CP value, if the distance is within a
+// predefined threshold.
+double
+OldFieldReorderAnalyzer::calculateCloseProximity(FieldNumType Field1,
+                                                 FieldNumType Field2) const {
+  if (Field1 == Field2)
+    // In old implementation, actually it's good to have a self-related field
+    // But we ignore this
+    return 0;
+  auto *Pair = CPBuilder->getCloseProximityPair(Field1, Field2);
+  if (Pair->first < 1e-3)
+    return 0;
+  if (Pair->second > DistanceThreshold)
+    return 0;
+  return Pair->first;
+}
+
+// Function to calculate FanOut of a field. Here, FanOut means the sum of
+// CP values that this field can generate with all the other remaining fields.
+// It's used to measure a potential benefit of bringing this field into NewOrder
+// and it can provide more opportunites to bring other fields.
+double OldFieldReorderAnalyzer::calculateFanOut(FieldNumType FieldNum) const {
+  double CPSum = 0;
+  for (auto &F : FieldsToTransform) {
+    if (F != FieldNum) {
```

Comment:
Remove brace on if. You can also do on the for, but keep it if you like it better that way.